### PR TITLE
Standalone Lite: DDP Spawn Strategy Family

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Adds the DDP **Spawn** family of strategies supported in Lite. 
Very similar to #14670 , but using the spawn launchers. 

Note: In the future, the spawn strategies will merge together with their non-spawn versions, as most logic related to process creation has been factored out to the launchers already. 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [ ] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

**Review by justusschock — APPROVED** [review]
> Note: In the future, the spawn strategies will merge together with their non-spawn versions, as most logic related to process creation has been factored out to the launchers already.

Already create an issue for this to keep track?

**Comment by awaelchli** [comment]
@justusschock, I created one: #14707

File: src/lightning_lite/strategies/tpu_spawn.py [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+if _TPU_AVAILABLE:
+    import torch_xla.core.xla_env_vars as xenv
+    import torch_xla.core.xla_model as xm
+    from torch_xla.core.xla_model import rendezvous
+    from torch_xla.distributed.parallel_loader import MpDeviceLoader
+else:
+    xm, xmp, MpDeviceLoader, rendezvous = [None] * 4
+
+
+class TPUSpawnStrategy(DDPSpawnStrategy):
```
Comment (awaelchli) [inline src/lightning_lite/strategies/tpu_spawn.py:47]: Let's consider naming this just TPUStrategy?

In the future, it will not subclass DDPSpawn strategy, and it has it's own launcher implementation anyway. Plus, the implementation differs quite significantly from DDPSpawnStrategy

File: src/lightning_lite/strategies/ddp_spawn.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+            self._process_group_backend,
+            self.global_rank,
+            self.world_size,
+            timeout=self._timeout,
+        )
+
+    def _get_process_group_backend(self) -> str:
+        return (
+            self._process_group_backend
+            or _get_process_group_backend_from_env()
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp_spawn.py:211]: This should be removed from here, as it'll be removed in PL too: https://github.com/Lightning-AI/lightning/pull/14693

File: src/lightning_lite/strategies/ddp_spawn.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+from lightning_lite.utilities.distributed import (
+    _get_process_group_backend_from_env,
+    distributed_available,
+    get_default_process_group_backend_for_device,
+)
+from lightning_lite.utilities.distributed import group as _group
+from lightning_lite.utilities.distributed import init_dist_connection, ReduceOp, sync_ddp_if_available
+from lightning_lite.utilities.rank_zero import rank_zero_only
+
+_DDP_FORK_ALIASES = (
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp_spawn.py:41]: Where is this used? Can't find it

File: src/lightning_lite/strategies/ddp_spawn.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    def _configure_launcher(self) -> None:
+        self._launcher = _MultiProcessingLauncher(self, start_method=self._start_method)
+
+    def setup_module(self, module: Module) -> Module:
+        return DistributedDataParallel(module=module, device_ids=self._determine_ddp_device_ids(), **self._ddp_kwargs)
+
+    def module_to_device(self, module: Module) -> None:
+        if self.root_device.type == "cuda":
+            # TODO(lite): This should be handled outside module_to_device, by a call to accelertor.setup_device()
```
Comment (carmocca) [inline src/lightning_lite/strategies/ddp_spawn.py:118]:
```suggestion
            # TODO(lite): This should be handled outside module_to_device, by a call to accelerator.setup_device()
```

File: src/lightning_lite/strategies/tpu_spawn.py (hunk 1 above)
Comment (carmocca) [reply src/lightning_lite/strategies/tpu_spawn.py:47]: I agree, but would prefer XLAStrategy. Having TPU in the name is confusing since TPU is an accelerator, not a strategy

**Review by carmocca — COMMENTED** [review]
I'll review again after https://github.com/Lightning-AI/lightning/pull/14670 is merged, since there's some overlap

**Review by Borda — APPROVED** [review]

File: tests/tests_lite/utilities/test_imports.py [hunk 5]
```diff
@@ -27,9 +27,9 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 
+from lightning_lite.strategies.fairscale import _FAIRSCALE_AVAILABLE
```
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/tests_lite/utilities/test_imports.py:30]: Okay. Do you know what the test_imports.py file is for?

File: src/lightning_lite/strategies/ddp_spawn.py (hunk 3 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/ddp_spawn.py:41]: It will be used in the accelerator connector, once it arrives #14692, for validation purposes.

File: tests/tests_lite/utilities/test_imports.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (carmocca) [reply tests/tests_lite/utilities/test_imports.py:30]: My comment above is incorrect. I thought this was something else. Sorry!

`test_imports` was added because we had an issue where `*_AVAILABLE` returned False, but the library could be imported (or was it vice versa?)

In any case, this test should be in either PL or Lite, but not both

File: src/lightning_lite/strategies/fairscale.py [hunk 6]
```diff
@@ -120,3 +121,56 @@ def _reinit_optimizers_with_oss(self, optimizers: List[Optimizer]) -> List["OSS"
...
+        strategy_registry.register(
+            "ddp_sharded_spawn_find_unused_parameters_false",
+            cls,
+            description="DDP Spawn Sharded Strategy with `find_unused_parameters` as False",
+            find_unused_parameters=False,
+        )
+        strategy_registry.register(
+            "ddp_sharded_spawn",
+            cls,
+            description=f"{cls.__class__.__name__}",
```
Comment (carmocca) [inline src/lightning_lite/strategies/fairscale.py:166]: Got it right this time

```suggestion
            description=cls.__class__.__name__,
```

File: src/lightning_lite/strategies/fairscale.py [hunk 7]
```diff
@@ -120,3 +121,56 @@ def _reinit_optimizers_with_oss(self, optimizers: List[Optimizer]) -> List["OSS"
...
+        )
+
+    def _reinit_optimizers_with_oss(self, optimizers: List[Optimizer]) -> List["OSS"]:
```
Comment (carmocca) [inline src/lightning_lite/strategies/fairscale.py:169]: This implementation should exactly match the one in `DDSharded`, no? Did they accidentally diverge in PL?

File: src/lightning_lite/strategies/fairscale.py [hunk 8]
```diff
@@ -120,3 +121,56 @@ def _reinit_optimizers_with_oss(self, optimizers: List[Optimizer]) -> List["OSS"
                 optimizers[x] = zero_optimizer
                 del optimizer
         return optimizers
+
+
+class DDPSpawnShardedStrategy(DDPSpawnStrategy):
+    """Optimizer sharded training provided by FairScale."""
```
Comment (carmocca) [inline src/lightning_lite/strategies/fairscale.py:127]: For consistency

```suggestion
    """Optimizer and gradient sharded training provided by FairScale with Spawn."""
```

**Review by carmocca — APPROVED** [review]

File: src/lightning_lite/strategies/fairscale.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/fairscale.py:169]: Yes, they seem to differ only in the multi-node aspect. While ddp spawn technically does support multi-node, it is not as convenient when launching processes, and hence I guess the original author(s) thought the regular ddp implementation should be the one with the multi-node optimizations. 

I'll think about what to do here.

File: tests/tests_lite/utilities/test_imports.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (awaelchli) [reply tests/tests_lite/utilities/test_imports.py]: Got it. Following up on this by importing _FAIRSCALE_AVAILABLE in PL from Lite and then move the test accordingly.

File: src/lightning_lite/strategies/fairscale.py (hunk 7 above)
Comment (awaelchli) [reply src/lightning_lite/strategies/fairscale.py:183]: Made them identical in [185dd64](https://github.com/Lightning-AI/lightning/pull/14675/commits/185dd6451744ef96b2cd1e34cf2bc162d08cf782)

Awaiting further reduction by #14707

File: src/lightning_lite/strategies/launchers/xla.py [hunk 9]
```diff
@@ -86,8 +86,7 @@ def _wrapping_function(
         return_queue: SimpleQueue,
         global_states: Optional[_GlobalStateSnapshot] = None,
     ) -> None:
-        # TODO(lite): Update worker setup once TPUSpawn strategy is in Lite
-        self._strategy._worker_setup(process_idx)
+        self._strategy._local_rank = process_idx
```
Comment (carmocca) [inline src/lightning_lite/strategies/launchers/xla.py:89]: @awaelchli I believe this is not correct, as this no longer sets `XLAStrategy._launched=True` and then #14926 fails with

```python
  File "/home/runner/work/lightning/tests/tests_lite/strategies/test_xla.py", line 17, in broadcast_on_tpu_fn
    result = strategy.broadcast(obj)
  File "/home/runner/work/lightning/src/lightning_lite/strategies/xla.py", line 146, in broadcast
    data_tensor = torch.tensor(data, device=self.root_device, dtype=torch.float)
  File "/home/runner/work/lightning/src/lightning_lite/strategies/xla.py", line 72, in root_device
    raise RuntimeError("Accessing the XLA device before processes have spawned is not allowed.")
```

I'm a bit confused about what's the best way to do this. 

Do `strategy.setup_environment()`?

If yes:
  - This does not set the local rank. Should it? or do we still manually set the local rank?
  - Should this also be done for the other Lite launchers?

Also, why did Lite remove `_worker_setup`? The logic being different between PL and Lite strategies is confusing.

This blocks #14926

Comment (awaelchli) [reply src/lightning_lite/strategies/launchers/xla.py:89]: This was done in reaction to your comment: https://github.com/Lightning-AI/lightning/pull/11073#issuecomment-1247355717
I think it is correct, otherwise many many tests for ddp spawn would fail, and tpu spawn is not fundamentally different regarding this local rank business. 

I commented on https://github.com/Lightning-AI/lightning/pull/14926 that maybe all that is missing in the test is a strategy.setup_environment().
